storage/innobase/row/row0mysql.cc

```diff
@@ -2826,6 +2842,170 @@ row_add_table_to_background_drop_list(table_id_t table_id)
 	return added;
 }
 
+#define MAX_TRUNCATE_FAIL_COUNT 5
+/** The master thread in srv0srv.cc calls this regularly to truncate files which
+we must delete in background after DROP TABLE have ended. Such lazy dropping of files
+is needed to alleviate I/O and dict_sys->mutex bottleneck of DROP TABLE.
+@return false if all files are deleted */
+bool
+row_truncate_file_for_mysql_in_background(void)
+{
+	if (row_mysql_truncate_list_inited == FALSE)
+		return false;
+
+	row_mysql_truncate_t* trun;
+
+	mutex_enter(&row_truncate_list_mutex);
+	trun = UT_LIST_GET_FIRST(row_mysql_truncate_list);
+	mutex_exit(&row_truncate_list_mutex);
+
+	if (trun == NULL) {
+		/* All files dropped */
+		return false;
+	}
+
+	DBUG_EXECUTE_IF("ib_before_row_truncate_file", os_thread_sleep(3000000););
+	struct stat st;
+	int ret;
+	ret = stat(trun->file_name, &st);
+	if (ret == -1) {
+		if (errno == ENOENT)
+			/* If for some reason the file has already been dropped
+			through some other mechanism, do not try to drop it */
+			goto already_dropped;
+		else {
+			trun->fail_count++;
+			if (trun->fail_count > MAX_TRUNCATE_FAIL_COUNT)
+				goto already_dropped; /* no more retry on this file */
+			else
+				return true; /* master thread would retry */
+		}
+	}
+
+	if ((long)st.st_size <= (long)srv_async_truncate_size*1024*1024) {
+		unlink(trun->file_name);
+	}
+	else {
+		truncate(trun->file_name, st.st_size - srv_async_truncate_size*1024*1024);
```

Comment:
I think that we should use `ftruncate()` on the file handle that we would keep open, to let the original `unlink()` proceed fast.